File: pr1-java/src/main/java/repository/database/GameDbRepository.java
package repository.database;

import domain.Game;
import domain.exceptions.DuplicateException;
import domain.exceptions.NotFoundException;
import repository.GameRepository;
import utils.Configuration;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class GameDbRepository implements GameRepository {
    public GameDbRepository() {
        Configuration.logger.info("Initializing GameDbRepository with properties: {} ", Configuration.properties);
    }

    public static Game getGameFromResultSet(ResultSet resultSet) throws SQLException {
        Integer gameId = resultSet.getInt("gameId");
        String name = resultSet.getString("name");
        String homeTeam = resultSet.getString("homeTeam");
        String awayTeam = resultSet.getString("awayTeam");
        Integer availableSeats = resultSet.getInt("availableSeats");
        Integer seatCost = resultSet.getInt("seatCost");
        return new Game(gameId, name, homeTeam, awayTeam, availableSeats, seatCost);
    }

    @Override
    public Iterable<Game> getAll() {
        Configuration.logger.traceEntry();

        Connection connection = DbUtils.getConnection();
        List<Game> games = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from games;")) {
            getGamesFromDatabase(games, preparedStatement);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(games);
        return games;
    }

    @Override
    public Game getOne(Integer id) throws NotFoundException {
        Configuration.logger.traceEntry();

        Connection connection = DbUtils.getConnection();
        try (ResultSet result = connection.prepareStatement("select * from games where gameId = " + id + ";").executeQuery()) {
            Game game = getGameFromResultSet(result);

            Configuration.logger.traceExit(game);
            return game;
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
            throw new NotFoundException();
        }
    }

    @Override
    public Game add(Game game) throws DuplicateException {
        Configuration.logger.traceEntry("Saving game {} ", game);

        try {
            getOne(game.getId());
            throw new DuplicateException();
        } catch (NotFoundException ignored) {
        }

        Connection connection = DbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("insert into games(gameId, name, homeTeam, awayTeam, availableSeats, seatCost) " + "values(?,?,?,?,?,?);")) {
            preparedStatement.setInt(1, game.getId());
            preparedStatement.setString(2, game.getName());
            preparedStatement.setString(3, game.getHomeTeam());
            preparedStatement.setString(4, game.getAwayTeam());
            preparedStatement.setInt(5, game.getAvailableSeats());
            preparedStatement.setInt(6, game.getSeatCost());

            int result = preparedStatement.executeUpdate();
            Configuration.logger.trace("Saved {} instances", result);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(game);
        return game;
    }

    @Override
    public Game remove(Integer id) throws NotFoundException {
        Configuration.logger.traceEntry("Removing game with id {} ", id);

        Game game = getOne(id);
        Connection connection = DbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("delete from games where gameId = " + id + ";")) {
            int result = preparedStatement.executeUpdate();
            Configuration.logger.trace("Deleted {} instances", result);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
            throw new NotFoundException();
        }

        Configuration.logger.traceExit(game);
        return game;
    }

    @Override
    public Game modify(Integer id, Game newGame) throws NotFoundException {
        Configuration.logger.traceEntry("Modifying game with id {} ", id);

        Game oldGame = getOne(id);
        Connection connection = DbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("update games set " +
                "name = '" + newGame.getName() +
                "', homeTeam = '" + newGame.getHomeTeam() +
                "', awayTeam = '" + newGame.getAwayTeam() +
                "', availableSeats = " + newGame.getAvailableSeats() +
                ", seatCost = " + newGame.getSeatCost() +
                " where gameId = " + id + ";")) {
            int result = preparedStatement.executeUpdate();
            Configuration.logger.trace("Modified {} instances", result);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(oldGame);
        return oldGame;

    }

    @Override
    public Iterable<Game> getGamesOrderedByAvailableSeats(Boolean reverse) {
        Configuration.logger.traceEntry();

        Connection connection = DbUtils.getConnection();
        List<Game> games = new ArrayList<>();
        String statement = "select * from games order by availableSeats";
        statement += reverse ? " desc;" : ";";
        try (PreparedStatement preparedStatement = connection.prepareStatement(statement)) {
            getGamesFromDatabase(games, preparedStatement);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(games);
        return games;
    }

    private void getGamesFromDatabase(List<Game> games, PreparedStatement preparedStatement) throws SQLException {
        try (ResultSet result = preparedStatement.executeQuery()) {
            while (result.next()) {
                Game game = getGameFromResultSet(result);
                games.add(game);
            }
        }
    }

    @Override
    public String toString() {
        StringBuilder string = new StringBuilder();
        for (Game game : getAll())
            string.append(game).append("\n");
        return string.toString();
    }
}

File: pr1-cs/Persistence/Database/GameDbRepository.cs
using System.Collections.Generic;
using System.Data;
using System.Linq;
using Model;
using Model.Exceptions;
using DuplicateException = Model.Exceptions.Thrift.DuplicateException;
using NotFoundException = Model.Exceptions.Thrift.NotFoundException;

namespace Persistence.Database
{
    public class GameDbRepository : IGameRepository
    {
        public IEnumerable<Game> GetAll()
        {
            var games = new List<Game>();
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "select * from games;";
            GetGamesFromDatabase(command, games);
            return games;
        }

        public Game GetOne(int id)
        {
            Game game;
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "select * from games where gameId = " + id + ";";
            using var dataReader = command.ExecuteReader();
            if (dataReader.Read())
                game = GetGameFromDataReader(dataReader);
            else
                throw new NotFoundException();
            return game;
        }

        public Game Add(Game game)
        {
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText =
                "insert into games(gameId, homeTeam, awayTeam, availableSeats, name, seatCost) values(@gameId, @homeTeam, @awayTeam, @availableSeats, @name, @seatCost);";

            var dataParameter = command.CreateParameter();
            dataParameter.ParameterName = "@gameId";
            dataParameter.Value = game.Id;
            command.Parameters.Add(dataParameter);

            dataParameter = command.CreateParameter();
            dataParameter.ParameterName = "@name";
            dataParameter.Value = game.Name;
            command.Parameters.Add(dataParameter);

            dataParameter = command.CreateParameter();
            dataParameter.ParameterName = "@homeTeam";
            dataParameter.Value = game.HomeTeam;
            command.Parameters.Add(dataParameter);

            dataParameter = command.CreateParameter();
            dataParameter.ParameterName = "@awayTeam";
            dataParameter.Value = game.AwayTeam;
            command.Parameters.Add(dataParameter);

            dataParameter = command.CreateParameter();
            dataParameter.ParameterName = "@availableSeats";
            dataParameter.Value = game.AvailableSeats;
            command.Parameters.Add(dataParameter);

            dataParameter = command.CreateParameter();
            dataParameter.ParameterName = "@seatCost";
            dataParameter.Value = game.SeatCost;
            command.Parameters.Add(dataParameter);

            var result = command.ExecuteNonQuery();
            if (result == 0)
                throw new DuplicateException();
            return game;
        }

        public Game Remove(int id)
        {
            var game = GetOne(id);
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "delete from games where gameId = " + id + ";";
            var result = command.ExecuteNonQuery();
            if (result == 0)
                throw new NotFoundException();
            return game;
        }

        public Game Modify(int id, Game newGame)
        {
            var oldGame = GetOne(id);
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "update games set " +
                                  "name = '" + newGame.Name +
                                  "', homeTeam = '" + newGame.HomeTeam +
                                  "', awayTeam = '" + newGame.AwayTeam +
                                  "', availableSeats = " + newGame.AvailableSeats +
                                  ", seatCost = " + newGame.SeatCost +
                                  " where gameId = " + id + ";";
            var result = command.ExecuteNonQuery();
            if (result == 0)
                throw new NotFoundException();
            return oldGame;
        }

        public IEnumerable<Game> GetGamesByAvailableSeatsDescending(bool reverse)
        {
            var games = new List<Game>();
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "select * from games order by availableSeats";
            command.CommandText += reverse ? " desc;" : ";";
            using var dataReader = command.ExecuteReader();
            while (dataReader.Read()) games.Add(GetGameFromDataReader(dataReader));

            return games;
        }

        public Game SetGameAvailableSeats(int id, int availableSeats)
        {
            var game = GetOne(id);
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "update games set availableSeats = " + availableSeats + " where gameId = " + id + ";";
            var result = command.ExecuteNonQuery();
            if (result == 0)
                throw new NotFoundException();
            return game;
        }

        private void GetGamesFromDatabase(IDbCommand command, List<Game> games)
        {
            using var dataReader = command.ExecuteReader();
            while (dataReader.Read()) games.Add(GetGameFromDataReader(dataReader));
        }

        public static Game GetGameFromDataReader(IDataReader dataReader)
        {
            var gameId = dataReader.GetInt32(dataReader.GetOrdinal("gameId"));
            var name = dataReader.GetString(dataReader.GetOrdinal("name"));
            var homeTeam = dataReader.GetString(dataReader.GetOrdinal("homeTeam"));
            var awayTeam = dataReader.GetString(dataReader.GetOrdinal("awayTeam"));
            var availableSeats = dataReader.GetInt32(dataReader.GetOrdinal("availableSeats"));
            var seatCost = dataReader.GetInt32(dataReader.GetOrdinal("seatCost"));

            return new Game(gameId, name, homeTeam, awayTeam, availableSeats, seatCost);
        }

        public override string ToString()
        {
            return GetAll().Aggregate("", (current, game) => current + (game + "\n"));
        }
    }
}